doc/sphinx/proof-engine/ssreflect-proof-language.rst

```diff
@@ -3636,6 +3636,231 @@ rewriting rule is stated using Leibniz equality (as opposed to setoid
+practical issues during the development of the corresponding scripts,
+notably as we might be forced to provide the rewrite tactic with
+complete terms, as shown by the simple example below.
+
+.. example::
+
+   .. coqtop:: reset none
+
+      From Coq Require Import ssreflect.
+      Set Implicit Arguments.
+      Unset Strict Implicit.
+      Unset Printing Implicit Defensive.
+
+   .. coqtop:: in
+
+      Axiom subnn : forall n : nat, n - n = 0.
+      Parameter G : (nat -> nat) -> nat -> nat.
+      Axiom eq_G :
+        forall F1 F2 : nat -> nat,
+        (forall n : nat, F1 n = F2 n) ->
+        forall n : nat, G F1 n = G F2 n.
+
+   .. coqtop:: all
+
+      Lemma example_G (n : nat) : G (fun n => n - n) n >= 0.
+
+   In this context, one cannot directly use ``eq_G``:
+
+   .. coqtop:: all fail
+
+      rewrite eq_G.
+
+   as we need to explicitly provide the non-inferable argument ``F2``,
+   which corresponds here to the term we want to obtain *after* the
+   rewriting step. We can thus do:
+
+   .. coqtop:: all abort
+
+      rewrite (@eq_G _ (fun _ : nat => 0)); last first.
+        by move=> m; rewrite subnn.
+
+   Otherwise, a more convenient approach consists in relying on the
+   :tacn:`under` tactic that alleviates this bookkeeping:
+
+   .. coqtop:: all abort
+
+      Lemma example_G (n : nat) : G (fun n => n - n) n >= 0.
+      under m: eq_G by rewrite subnn.
+
+  ``(* FIXME: replace with do... *)``
+
+.. tacn:: under {? @occ_switch} {? [ @r_pattern ] } @term {? => @i_item} {? do ( @tactic | [ {*| @tactic } ] ) }
+   :name: under
+
+    Rewrite under :token:`term`. We distinguish two modes,
+    :ref:`interactive mode <under_interactive>` without a ``do`` clause, and
+    :ref:`one-liner mode <under_one_liner>` with a ``do`` clause,
+    which are explained in more detail below.
+
+   .. exn:: Incorrect number of tactics (expected N tactics, was given N').
+
+      This error can occur when using the version with a ``do`` clause.
+
+.. _under_interactive:
+
+Interactive mode
+````````````````
+The execution of the Ltac expression:
+:n:`under {@occ_switch}[@r_pattern]@term => [i|j].`
+involves the following steps:
+
++ It performs a full-blown :n:`rewrite {@occ_switch}[@r_pattern]@term`,
+  without failing like in the above example with ``eq_G``, but
+  creating evars.
++ As a 2-branches intro-pattern is provided, it checks that 3 subgoals
+  have been created (the last one being the initial, main subgoal that
+  now contains some evars).
++ In the first 2 subgoals, it introduces ``i``, resp. ``j``, and tests
+  whether these 2 subgoals are some (quantified) equalities involving
+  an evar. For example: ``some_expr(i) = ?Goal i``.
++ If this is the case, the subgoal is manipulated to get the
+  provably-equivalent goal ``Under _ (some_expr i) (?Goal i)``,
+  pretty-printed as ``Under[ some_expr i ]``.
+  This change permits to avoid instantiating the evar by accident
```

Comment:
"permits to avoid" is somewhere between awkward and ungrammatical.  "permits avoiding" is the usual wording.  But then "permits avoiding instantiating" is not so graceful, either.  Maybe just use "This change avoids instantiating" or "This change lets you avoid instantiating",.